### PR TITLE
Added experimental.pixelShaderImagePath

### PR DESCRIPTION
**Pull request by mrange**
I realize I might be one of the few developers that care about custom
shader support in terminal but I thought it's worth proposing it and see
what you think.

This is to support custom shaders with custom textures.

I was thinking of exposing the background image to the shader but that
felt complicated after looking into it.

I have tested exploratively. I think the texture loader is possible to
unit test so that is a possible improvement.

The error reporting (as with other custom pixel shader code) is not very
good. That is also an area that I could improve upon.

I do think the risk of adding this is rather low as the new code is only
executed when experimental.pixelShaderImagePath is set.

### Details

Only added to the Atlas engine. Unsure if needs adding in DXEngine?

Instead I load the texture using WIC into a shader resource view. When
binding shader resources I test for presence of custom texture and bind
it to register t1.

The image loading code was found in [the D3D Texture documentation].
It's a mouthful but seems rather robust.

Tested setting: "experimental.pixelShaderImagePath"

1. Tested not specifying it.
2. Tested setting it.
3. Tested changing it (the changes are picked up)
4. Tested invalid path
5. Tested a custom shader that made use of the custom texture.
6. 
[the D3D Texture documentation]: https://learn.microsoft.com/en-us/windows/win32/direct3d11/overviews-direct3d-11-resources-textures-how-to

### CODE REVIEW

**Comment by mrange** [comment]
image

The blazor logo is a custom shader image.

**Comment by mrange** [comment]
I get a bunch of spell check issues in code I haven't modified. What should I do about them?

**Comment by mrange** [comment]
Finally all checks are green

File: src/renderer/atlas/AtlasEngine.cpp [hunk 1]
```diff
@@ -27,6 +28,559 @@
 
 using namespace Microsoft::Console::Render;
 
+namespace
+{
+    // From: https://learn.microsoft.com/en-us/windows/win32/direct3d11/overviews-direct3d-11-resources-textures-how-to
```
Comment (lhecker) [inline src/renderer/atlas/AtlasEngine.cpp:33]: This code might be outdated. It's actually taken from here which is also the code we might want to copy if anything: <https://github.com/microsoft/DirectXTK/blob/2530d168f50ecca48bf4f2f7f22a59f840939816/Src/WICTextureLoader.cpp>

I noticed that you renamed/merged some functions which I'd prefer if that wasn't the case, if I'm honest. Simply because those original function names will make finding and comparing this code with the original source of this code easier.

File: src/renderer/atlas/AtlasEngine.cpp [hunk 2]
```diff
@@ -676,6 +1230,18 @@ void AtlasEngine::_createResources()
...
+        {
+            _r.customShaderTexture = LoadCustomShaderTexture(
+                _r.device.get(),
+                _r.deviceContext.get(),
+                _api.customPixelShaderImagePath);
+        }
+        else
+        {
+            _r.customShaderTexture = CustomShaderTexture();
+        }
```
Comment (lhecker) [inline src/renderer/atlas/AtlasEngine.cpp:1243]: You can drop this else branch. `_r` will always be emptied initially.

File: src/renderer/atlas/AtlasEngine.r.cpp [hunk 3]
```diff
@@ -243,7 +243,18 @@ void AtlasEngine::_renderWithCustomShader() const
...
         _r.deviceContext->PSSetShader(_r.customPixelShader.get(), nullptr, 0);
         _r.deviceContext->PSSetConstantBuffers(0, 1, _r.customShaderConstantBuffer.addressof());
-        _r.deviceContext->PSSetShaderResources(0, 1, _r.customOffscreenTextureView.addressof());
+        if (_r.customShaderTexture.TextureView)
+        {
+            ID3D11ShaderResourceView* const resourceViews[]{
+                _r.customOffscreenTextureView.get(),
+                _r.customShaderTexture.TextureView.get(),
+            };
+            _r.deviceContext->PSSetShaderResources(0, 2, static_cast<ID3D11ShaderResourceView* const*>(resourceViews));
```
Comment (lhecker) [inline src/renderer/atlas/AtlasEngine.r.cpp:252]: You could deduplicate this code by making the length dependent on the existence of the texture view like so:
```cpp
ID3D11ShaderResourceView* const views[]{
    _r.customOffscreenTextureView.get(),
    _r.customShaderTexture.TextureView.get(),
};
const UINT numViews = _r.customShaderTexture.TextureView ? 2 : 1;
_r.deviceContext->PSSetShaderResources(0, numViews, &views[0]);
```

File: src/renderer/dx/DxRenderer.cpp [hunk 4]
```diff
@@ -1031,6 +1037,18 @@ try
...
+{
+    if (_pixelShaderImagePath != value)
+    {
+        _pixelShaderImagePath = std::wstring{ value };
+        _recreateDeviceRequested = true;
+        LOG_IF_FAILED(InvalidateAll());
+    }
+}
+CATCH_LOG()
+
```
Comment (lhecker) [inline src/renderer/dx/DxRenderer.cpp:1051]: You can revert the changes to `src/renderer/dx` as it doesn't use `_pixelShaderImagePath`.

**Review by lhecker — COMMENTED** [review]
I'm generally fine with this, because while we might remove this option in the future to replace it with something more powerful I believe that texture loading like this will at least remain.

I just have a few concerns...

**Review by lhecker — APPROVED** [review]

**Comment by DHowett** [comment]
_Exciting!_

**Comment by mrange** [comment]
Thanks for the review. I am a little bit short in time during weekdays, but I think I can address some of your concerns over the weekend.

**Comment by WSLUser** [comment]
I have to confess as a user, I'm unsure what the difference between this and the already existing shader support is. I don't think we should have 2 settings regarding shaders. That'll be very confusing. If there's some type of support missing from the other shader code, could you simply merge the 2 instead so we only have a single user facing option to specify a shader file?

**Comment by mrange** [comment]
@WSLUser  I don't mind merging the options but this setting is to support loading a shader texture like a PNG that the shader source code can make use of. I am testing with these settings:
```
        "defaults": 
        {
            "experimental.pixelShaderImagePath": "C:\\temp\\cake.png",
            "experimental.pixelShaderPath": "C:\\code\\github\\windows-terminal-shaders\\cake.hlsl"
        },
```
So the `pixelShaderImagePath` points to a PNG loaded as a texture which the shader pointed out by `pixelShaderPath` uses.

So in my example it looks like this:

image

The background is computed by the shader but cake logo comes from the PNG which the shader overlays over the background.

How would you prefer the configuration option to look like?

**Comment by WSLUser** [comment]
Ok I think I see what you mean. That makes more sense but this will definitely need some documentation to avoid the confusion I had. In this instance I see that 2 settings will in fact be needed but they should be grouped together as this one requires the other one to be set (so the shader path should be set as a hard dependency before trying to fetch a ShaderImage path). I think you might be able to use the json schema for this, which will need updating regardless.

File: src/renderer/dx/DxRenderer.cpp (hunk 4 above)
Comment (mrange) [reply src/renderer/dx/DxRenderer.cpp:1051]: I think I need to implement the interface? But it can be implemented as a NOP?

File: src/renderer/atlas/AtlasEngine.cpp (hunk 2 above)
Comment (mrange) [reply src/renderer/atlas/AtlasEngine.cpp:1243]: Done

File: src/renderer/atlas/AtlasEngine.r.cpp (hunk 3 above)
Comment (mrange) [reply src/renderer/atlas/AtlasEngine.r.cpp:252]: Done

File: src/renderer/atlas/AtlasEngine.cpp (hunk 1 above)
Comment (mrange) [reply src/renderer/atlas/AtlasEngine.cpp:33]: In progress...

**Comment by mrange** [comment]
@lhecker in response to your comment on looking at DirectXTK for texture loading code I made the change that I added that repo as a submodule. I noticed you had submodules already so I figured it wouldn't be too much of a paradigm shift to add another one. Not sure this is a how you like it done but this is my current proposal.

If we are to go with the submodule I would like some advice on best way on how to include the source code from it. The approach taken works it seems but is it a good enough solution?

**Comment by mrange** [comment]
So because of the "controversial" addition of a submodule I consider this PR to be a proposal with need of more testing cleanup to be merged.

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 5]
```diff
@@ -1716,7 +1716,7 @@ til::point Terminal::GetViewportRelativeCursorPosition() const noexcept
 // These functions are used by TerminalInput, which must build in conhost
 // against OneCore compatible signatures. See the definitions in
 // VtApiRedirection.hpp (which we cannot include cross-project.)
-// Since we do nto run on OneCore, we can dispense with the compatibility
```
Comment (mrange) [inline src/cascadia/TerminalCore/Terminal.cpp:1719]: Not sure I capture the intent of the original comment. The spell checker complains about nto and the only way I got the sentence to work was like in my change.

File: src/renderer/atlas/TextureLoader.h [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Microsoft::Console::Render
+{
+    struct ShaderTexture
+    {
+        wil::com_ptr<ID3D11Resource> Texture;
+        wil::com_ptr<ID3D11ShaderResourceView> TextureView;
+    };
+
+    Microsoft::Console::Render::ShaderTexture LoadShaderTextureFromFile(
```
Comment (mrange) [inline src/renderer/atlas/TextureLoader.h:14]: I put this in a separate file as I felt that could help if wanted to add the function to DxEngine as well? What is the status of DxEngine? How do I test it?

File: .github/actions/spelling/allow/allow.txt [hunk 7]
```diff
@@ -32,6 +32,7 @@ ftp
 fvar
 gantt
 gcc
+Geddy
```
Comment (mrange) [inline .github/actions/spelling/allow/allow.txt:35]: Extended the word lists to make the spell checker happier.

File: src/renderer/atlas/TextureLoader.cpp [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+            D3D11_USAGE_DEFAULT,
+            D3D11_BIND_SHADER_RESOURCE,
+            0,
+            0,
+            DirectX::WIC_LOADER_DEFAULT
+                // TODO: Should we ignore SRGB conversion?
+                //  If we use the default settings the texture is converted
+                //  from SRGB into linear RGB which do make a lot of sense but
+                //  can also be a somewhat surprising to devs not used to it.
+                | DirectX::WIC_LOADER_IGNORE_SRGB,
```
Comment (mrange) [inline src/renderer/atlas/TextureLoader.cpp:28]: Perhaps could be configurable if needed at some later point?

**Comment by mrange** [comment]
Bump. Still looking for feedback if the latest approach is acceptable (ie bring in DirectXTK as a submodule)

**Comment by lhecker** [comment]
@mrange I think it would've been fine if you just copied DirectXTK's source code into this project. But since it's just a submodule, it's easy to add/remove it and so I'm also fine with your new approach. I think this is good to be merged as is. 👍

File: src/renderer/atlas/TextureLoader.cpp (hunk 8 above)
Comment (lhecker) [reply src/renderer/atlas/TextureLoader.cpp:28]: Our render target is in uncorrected screen gamma, which means that if you were to not use `WIC_LOADER_IGNORE_SRGB`, it would result in no gamma correction at all. This way we at least get something close to it (luckily most monitors are close to sRGB's gamma).

File: src/renderer/dx/DxRenderer.cpp (hunk 4 above)
Comment (lhecker) [reply src/renderer/dx/DxRenderer.cpp:1051]: You don't need it, since you already provided a no-op implementation in `IRenderEngine`.

File: src/renderer/atlas/TextureLoader.h (hunk 6 above)
Comment (lhecker) [reply src/renderer/atlas/TextureLoader.h:14]: Our intention is to slowly replace DxEngine with AtlasEngine over time. As such we don't really plan to implement any improvements for it anymore unless they have a great usability impact for everyone. Personally, I wouldn't port this PR over to DxEngine.

File: src/renderer/atlas/atlas.vcxproj [hunk 9]
```diff
@@ -7,10 +7,12 @@
     <ProjectName>RendererAtlas</ProjectName>
     <TargetName>ConRenderAtlas</TargetName>
     <ConfigurationType>StaticLibrary</ConfigurationType>
+    <IncludePath>$(SolutionDir)\dep\DirectXTK\Inc;$(IncludePath)</IncludePath>
   </PropertyGroup>
   <Import Project="$(SolutionDir)src\common.build.pre.props" />
   <Import Project="$(SolutionDir)src\common.nugetversions.props" />
   <ItemGroup>
+    <ClCompile Include="..\..\..\dep\DirectXTK\Src\WICTextureLoader.cpp" />
```
Comment (mrange) [inline src/renderer/atlas/atlas.vcxproj:15]: Perhaps should use $(SolutionDir) as well.

File: src/renderer/atlas/pch.h [hunk 10]
```diff
@@ -25,6 +25,9 @@
 #include <dxgidebug.h>
 #include <VersionHelpers.h>
 
+#include <wrl/client.h>
```
Comment (mrange) [inline src/renderer/atlas/pch.h:28]: Required by TextureLoader.

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 5 above)
Comment (mrange) [reply src/cascadia/TerminalCore/Terminal.cpp:1719]: Should go into the spell check commit

File: src/renderer/atlas/TextureLoader.h (hunk 6 above)
Comment (mrange) [reply src/renderer/atlas/TextureLoader.h:14]: Ok

File: src/renderer/dx/DxRenderer.cpp (hunk 4 above)
Comment (mrange) [reply src/renderer/dx/DxRenderer.cpp:1051]: Ok

File: src/renderer/atlas/TextureLoader.cpp (hunk 8 above)
Comment (mrange) [reply src/renderer/atlas/TextureLoader.cpp:28]: Ok

File: src/renderer/atlas/TextureLoader.h (hunk 6 above)
Comment (mrange) [reply src/renderer/atlas/TextureLoader.h:14]: I might leave it as is as I don't think it hurts terribly having it in its own file and might help a unit test

File: src/renderer/dx/DxRenderer.cpp (hunk 4 above)
Comment (mrange) [reply src/renderer/dx/DxRenderer.cpp:1051]: Will change to NOPs

File: src/renderer/dx/DxRenderer.cpp [hunk 11]
```diff
@@ -1017,6 +1017,14 @@ std::wstring_view DxEngine::GetPixelShaderPath() noexcept
     return _pixelShaderPath;
 }
 
+std::wstring_view DxEngine::GetPixelShaderImagePath() noexcept
+{
+    // NOP implementation as DxEngine is to be replaced by Atlas
+    //  so recommended to not backport this function
+    static std::wstring empty;
```
Comment (mrange) [inline src/renderer/dx/DxRenderer.cpp:1024]: Magic statics works now? Or how should you create an empty wstring_view?

File: src/renderer/dx/DxRenderer.cpp (hunk 4 above)
Comment (mrange) [reply src/renderer/dx/DxRenderer.cpp:1047]: done

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 12]
```diff
@@ -1,17 +1,17 @@
 <?xml version="1.0" encoding="utf-8"?>
 <root>
-  <!-- 
-    Microsoft ResX Schema 
-    
+  <!--
```
Comment (mrange) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:3]: Bunch of WS changes when I just tried to fix a spelling error. I can revert it if you like

File: src/renderer/atlas/atlas.vcxproj [hunk 13]
```diff
@@ -7,10 +7,12 @@
     <ProjectName>RendererAtlas</ProjectName>
     <TargetName>ConRenderAtlas</TargetName>
     <ConfigurationType>StaticLibrary</ConfigurationType>
+    <IncludePath>$(SolutionDir)\dep\DirectXTK\Inc;$(IncludePath)</IncludePath>
   </PropertyGroup>
   <Import Project="$(SolutionDir)src\common.build.pre.props" />
   <Import Project="$(SolutionDir)src\common.nugetversions.props" />
   <ItemGroup>
+    <ClCompile Include=".$(SolutionDir)\dep\DirectXTK\Src\WICTextureLoader.cpp" />
```
Comment (lhecker) [inline src/renderer/atlas/atlas.vcxproj:15]: There's an extra dot here. I think this is breaking the build?

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 12 above)
Comment (lhecker) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:3]: Yeah they would need to be reverted before merging unfortunately.

**Review by lhecker — APPROVED** [review]

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 12 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:3]: This is one our old tooling problems. The resw files _must_ be saved with visual studio, otherwise they'll flip/flop their line endings and whitespace every commit that touches them. It's obnoxious.

Comment (mrange) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:3]: Will fix.

File: src/renderer/atlas/atlas.vcxproj (hunk 13 above)
Comment (mrange) [reply src/renderer/atlas/atlas.vcxproj:15]: Will fix. Been away for a week so slow response from me

**Comment by mrange** [comment]
Spellchecker still complains about: 
```
#### Unrecognized words (1)

otepad
```
But didn't find any `otepad` strings in git repo.

File: src/renderer/atlas/atlas.vcxproj (hunk 13 above)
Comment (mrange) [reply src/renderer/atlas/atlas.vcxproj]: Fixed. Let's see if it helps

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 12 above)
Comment (mrange) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw]: Fixed.

**Comment by DHowett** [comment]
(Thanks for being patient with us. I know that it's been some time. I wanted myself assigned so that I could vet the dependency and the risk we'd take on if we further brought it into Windows. We probably wouldn't be taking on any risk, but I just want to make sure.)

**Comment by mrange** [comment]
If you are concerned about bringing in the dependency into windows builds I could put some `#ifdef` in there

File: src/renderer/atlas/AtlasEngine.h [hunk 14]
```diff
@@ -1105,6 +1110,7 @@ namespace Microsoft::Console::Render
             bool enableTransparentBackground = false;
 
             std::wstring customPixelShaderPath; // changes are flagged as ApiInvalidations::Device
+            std::wstring customPixelShaderImagePath;
```
Comment (DHowett) [inline src/renderer/atlas/AtlasEngine.h:1113]: nit: i like how _each_ line says where changes are flagged; should we just propagate that here too?

```suggestion
            std::wstring customPixelShaderImagePath; // changes are flagged as ApiInvalidations::Device
```

File: src/renderer/atlas/TextureLoader.cpp (hunk 8 above)
Comment (DHowett) [reply src/renderer/atlas/TextureLoader.cpp:28]: Does that mean that this TODO is TODONE?

File: src/renderer/atlas/TextureLoader.h (hunk 6 above)
Comment (DHowett) [reply src/renderer/atlas/TextureLoader.h:14]: I agree with your intuition here @mrange -- I don't think DxEngine is dead yet, especially given that we have some issues with NewNewAtlasEngine that may prevent broad adoption . . . today. :smile:

File: src/renderer/atlas/atlas.vcxproj [hunk 15]
```diff
@@ -7,10 +7,12 @@
     <ProjectName>RendererAtlas</ProjectName>
     <TargetName>ConRenderAtlas</TargetName>
     <ConfigurationType>StaticLibrary</ConfigurationType>
+    <IncludePath>$(SolutionDir)\dep\DirectXTK\Inc;$(IncludePath)</IncludePath>
   </PropertyGroup>
   <Import Project="$(SolutionDir)src\common.build.pre.props" />
   <Import Project="$(SolutionDir)src\common.nugetversions.props" />
   <ItemGroup>
+    <ClCompile Include="$(SolutionDir)\dep\DirectXTK\Src\WICTextureLoader.cpp" />
```
Comment (DHowett) [inline src/renderer/atlas/atlas.vcxproj:15]: Alright, my only complaint is this one.

When I first saw the submodule, I was like ... hey okay we're bringing in a big library with a bunch of code to do a specific thing. My first instinct was to say, "Let's use a nuget package if one exists; that'll help us avoid bringing in too much of somebody else's code"

Now that it looks like it's just one file, I'm more inclined to say...

_Let's just copy it into `dep/DirectXTK` and give it a `cgmanifest` just like we did with `dep/jsoncpp`._ If it's really one file, it will be much easier for us to carry it around than a submodule, and if we choose to bring it into Windows we can move it from `dep` to `oss` (the Windows build system doesn't do submodules!)

**Review by DHowett — CHANGES_REQUESTED** [review]
I put this off because I was vaguely afraid of it; after all, the shader code is already hard to grok for the folks who aren't deep in the renderer. It turned out to be a freakin' breeze, however. I'm sorry I left you on read for so long.

My only other request, apart from the more direct one of "just copy the files we need", is: we should add a callout to the `$/NOTICES.md` file in the section for other microsoft projects we rely on. It's not a legal requirement, but since we're all part of a big open-source community it feels like the right thing to do.

Thanks again, and thanks so much for your patience.

**Comment by DHowett** [comment]
It's not exactly fair for me to take a while to respond and then let the bot call _you_ slow, so I'm poking it along to remove the "no recent activity" tag :)

**Comment by mrange** [comment]
Hope to get some time coming weekend to look into comments. The weekends before been buzy.

**Comment by zadjii-msft** [comment]
Hey we're trying to clean out our PR backlog: I'm gonna tag this up for discussion next team sync and figure out if we're bringing this in or not. The change Dustin called out above was simple enough that we could probably do it on your behalf, but Leonard had some other thoughts on another way we could go about this.... something about comments and shadertoy and I honestly didn't understand all he said 😝

**Comment by zadjii-msft** [comment]
To add more notes we said after I sent that: Maybe we just pull this in as-is, because it handles a lot of the plumbing for loading images and passing them down. THEN, as a follow up, we remove this setting, and replace it with the ShaderToy thing. That's the great thing with it being an experimental setting - we _can_ just remove it later if we have a better idea 😉 

(there's also probably some work that needs doing to update this for the latest version of the Atlas Engine (which went through an entire re-write since this PR was opened)

@mrange you are the Terminal shader master though, so let us know what you're thinking. Sorry for letting this stagnate so long! Let's get this in for 1.21 ☺️

**Comment by mrange** [comment]
Hi sorry for not answering. I have been travelling and doing shader presentations for .NET developers. Anyway, I will try to have an answer after work.

**Comment by mrange** [comment]
So I obviously like to see this merged and I think it will enable some cool things for us. The obvious thing is ofc an image like a logo but it will also enable importing noise images which are very useful things as it can be costly to do a compute for.

When I wrote it after I got the tip about using the external library I felt pretty confident that this was a rather safe change as it simplified the code quite much.

What is this shadertoy thing you are discussing?

**Comment by mrange** [comment]
Re-reading the PR. Most changes are just forwarding data and some spelling errors. The actual works is mostly handled by DirectXTk or rather simple (like checking if we have an extra texture and in that case include). 

I think personally it's rather low risk especially since this is new experimental setting but obviously you might disagree

**Comment by zadjii-msft** [comment]
> What is this shadertoy thing you are discussing?

I honestly don't know. @lhecker mentioned something about being able to pass a texture to the shader using a comment in the body of the shader itself, that shadertoy would then load as a texture for you? That was at the end of the day though and I probably misunderstood 😅

I can handle updating to main, and pulling in `DirectXTK` as a raw include if you'd like. That'll get this in and get us moving ☺️

**Comment by lhecker** [comment]
shadertoy has a very constrained template of things you can do in a shader: https://www.shadertoy.com/howto
...but it allows you to do some super impressive stuff: https://www.shadertoy.com/view/ssjyWc

I figured we could - over time - offer an API that is inspired by shadertoy's, but with the web interface configuration for the 4 channels and the buffer stages replaced with comments or similar that can be parsed during shader loading, or by having an additional .json file that contains such metadata.

I know that FX11 would fit quite well here, but it's not actively maintained, nor are there plans to port it to D3D12.

**Comment by zadjii-msft** [comment]
image

**Comment by mrange** [comment]
> shadertoy has a very constrained template of things you can do in a shader: https://www.shadertoy.com/howto ...but it allows you to do some super impressive stuff: https://www.shadertoy.com/view/ssjyWc
> 
> I figured we could - over time - offer an API that is inspired by shadertoy's, but with the web interface configuration for the 4 channels and the buffer stages replaced with comments or similar that can be parsed during shader loading, or by having an additional .json file that contains such metadata.
> 
> I know that FX11 would fit quite well here, but it's not actively maintained, nor are there plans to port it to D3D12.

I have used shadertoy quite much :) (322 public shaders). 

Not surprisingly I would like to see more shader capabilities. While multipasses (and retaining the previous frame pass) is great I personally would think I would love some kind of live editing before to simplify development.

Imagine for example that windows terminal reloads the shader code automatically when changed. If something goes wrong catch the error and show it as the shader texture.

That way I could sit and develop shaders real-time in VIM. That would be a very cool demo! And also helps productivity.

Also would like to control the pixelshader version. 

Then ofc capturing the windows sound and provide it as a texture with both the signal and the FFT opens up for cool stuff.

And so on :)

**Comment by mrange** [comment]
As for deploying shaders what about a folder with all the content in (also zip)?

In the folder images, Common.fx, BufferA.fx, BufferB.fx, BufferC.fx and Main.fx (only Main.fx required). Some kind of way to configure the input textures (not sure how).

**Comment by mrange** [comment]
Perhaps look at https://github.com/Gargaj/Bonzomatic for how they store "projects". Seems like pretty smart and supports DX

**Review by zadjii-msft — APPROVED** [review]
Yea I mean, I read most of this as I merged the code, and I'm cool with it. 

I'd suggest that @lhecker be the second, since atlas is his baby.

**Comment by mrange** [comment]
I read through the latest changes and at least to me they looked ok.

**Comment by lhecker** [comment]
Yeah, this PR looks good to me.

However, I'll try and see if I can write my own WIC loader, because I'd personally prefer if we minimized our external dependencies for now. I think it's unlikely we'll depend on other DirectXTK functionality in the future and so I'm not entirely sure if the ~40k line dependency is worth it, if I'm entirely honest. Reading through `WICTextureLoader.cpp` I get the feeling that we can write a significantly reduced version of it by relying on the fact that we only support recent Windows versions. If my own "WICTextureLoader" sucks, I'll approve the PR. 😅

**Comment by mrange** [comment]
@lhecker  I will just mention that IIRC my first PR had my own texture loader but I was then pointed DirectXTK  :)

**Comment by mrange** [comment]
@lhecker it's all good. I am just happy this gets in eventually :) I am planning to do some noise texturing stuff.

**Comment by lhecker** [comment]
After some thinking, I realized we don't need anything but `GUID_WICPixelFormat32bppPBGRA`. BGRA is the most widely supported native decoder format on Windows and is sufficient for our swap-chain (which is not sRGB and doesn't support HDR). I also felt like we want it to be **P**remultiplied so you don't have to do that yourself in the shader first. This makes all the lookup tables unnecessary and allowed me to shrink the entire WIC code down to <60 LOC.

I tested it with a HDR HEIF image, a 400MP JPEG, and a grayscale GIF, as well as a semi-transparent PNG and they all worked correctly.

**Review by lhecker — APPROVED** [review]

**Review by DHowett — APPROVED** [review]
Let's DO this

**Comment by mrange** [comment]
Cool when do you think this is available in a preview I can install?

**Comment by DHowett** [comment]
@mrange In about 9 hours! If you're a [canary user](https://github.com/microsoft/terminal/discussions/16121#discussion-5719075), that is :D
